app/lib/foreman_deployments/tasks/search_task_definition.rb

```diff
@@ -7,6 +7,26 @@ def run
           results = SerializableArray.new(results)
           SearchTaskDefinition.create_output(results, output)
         end
+
+        def self.dereference_output(storage, path)
```

Comment:
Let's begin with the fact that I think that we need to get rid of chain recording, if we want to allow custom output from a task. The reason for this is that if we record the chain, we are responsible to apply it, and this is not always trivial/possible.
The example you are presenting here is very interesting. First, I assume that the snippet belongs to an enveloping task (something like CreateClusterTask), to be more specific - to the planning phase of it.
Let's analyze this example to more details:

1. There is data transformation here: `host_creation.output[:mac]`
2. I see an unusual task ordering here - plan envelope task -> plan CreateHostTask -> **execute CreateHostTask**-> plan DhcpReserve task -> execute DhcpReserveTask
 
Let's start with [1]: here we can see a data transformation from the output of host creation to input of dhcp reservation. You have described a very simple transform but it doesn't have to be this way. What if the transformation needed here is something like `host_creation.output[:mac].to_binary_array`, or even in a more problematic case, the transformation is not defined on the object itself, something like this: `convert_to_msb_format(host_creation.output[:mac])`. I think the transformation is responsibility of the enveloping task in this example. We sure need to provide a helper that will make the code easier for the user, but recording doesn't solve this. I would suggest something like this:
``` Ruby
  plan_action(DhcpReserve, transform(host_creation) { |output| output.mac.to_binary_array } )
```
under the hood it will create a reference class that will save the delegate and apply it when the output will be ready.
This leads me to issue [2]: Since DhcpCreation task does not have assume some kind of dependency between tasks, it expects plain mac address. This data will be available only after CreateHost task will finish its execution phase. Basically that means that the task scheduling will be similar to [thread pool](https://msdn.microsoft.com/en-us/library/3dasc8as.aspx). The same goes for the enveloping task itself - let's say you want to plan different tasks based on some information of any other task: 
``` Ruby
  host_creation = plan_action(CreateHost, host_attrs)
  if host_creation.output[:status] == :success
    plan_action(DhcpReserve, host_creation.output[:mac])
  else
    plan_action(DeleteHost, host_attrs)
  end
```
You can continue the plan phase only after you have the result of the host creation.